File: src/go/partition/algo/mlt_dyn.go
package algo

import (
	"sync"
	"runtime"
	partition "golang/partition/lib"
)

func Mlt_dyn(arr [][]int) []int32 {
	wg := sync.WaitGroup{}
	nThreads := runtime.GOMAXPROCS(0)
	
	var result []int32 = make([]int32, len(arr))
	var partitionsChannel = make(chan partition.Task, nThreads)

	for i := 0; i < nThreads; i++ {
		wg.Add(1)

		go func(i int) {
			defer wg.Done()

			for task := range partitionsChannel {
				result[task.Index] = 0

				var combs int64 = 1 << (len(task.Array) - 1)
				var all int64 = (1 << len(task.Array)) - 1

				var problem_sum int = partition.Sum(task.Array, len(task.Array), all)
				if problem_sum%2 != 0 {
					continue
				}
				var half_sum int = problem_sum / 2

				for j := int64(0); j < combs; j++ {
					var sum int = partition.Sum(task.Array, len(task.Array), j)
					if sum == half_sum {
						result[task.Index] = 1
						break
					}
				}
			}
		}(i)
	}

	for i, problem := range arr {
		partitionsChannel <- partition.Task{Index: i, Array: problem}
	}
	close(partitionsChannel)

	wg.Wait()
	return result
}

File: src/julia/partition/algo/mlt_dyn.jl
module Mlt_dyn

include("../partition.jl")
using .Partition

function mlt_dyn(arr::Vector{Vector{Int64}})
    # an array of bools
    result = Vector{Bool}(undef, length(arr))

    Threads.@threads for i in 1:length(arr)
        result[i] = false
        size = length(arr[i])

        numOfCombinations = 1 << (size - 1)
        allNumbersMask = (1 << size) - 1

        problem_sum = partition_sum(arr[i], size, allNumbersMask)
        if problem_sum % 2 != 0
            continue
        end
        half_problem_sum = problem_sum / 2

        for j in 1:numOfCombinations
            sum = partition_sum(arr[i], size, j)
            if sum == half_problem_sum
                result[i] = true
                break
            end
        end
    end

    return result
end

end # module